File: other-programming-languages/c++/2024/2024-003/main.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>

using namespace std;

// Function to sum all mul(X, Y) operations
int sumMulOperations(const string& corruptedMemory) {
    regex pattern(R"(mul\((\d{1,3}),(\d{1,3})\))");
    smatch match;
    int total = 0;
    string::const_iterator searchStart(corruptedMemory.cbegin());

    // Find all matches and calculate the sum of products
    while (regex_search(searchStart, corruptedMemory.cend(), match, pattern)) {
        int num1 = stoi(match[1]);
        int num2 = stoi(match[2]);
        total += num1 * num2;
        searchStart = match.suffix().first;
    }

    return total;
}

// Function to sum only enabled mul(X, Y) operations
int sumEnabledMulOperations(const string& corruptedMemory) {
    regex pattern(R"((do\(\)|don't\(\)|mul\((\d{1,3}),(\d{1,3})\)))");
    smatch match;
    int totalSum = 0;
    bool mulEnabled = true; // mul instructions are enabled at the start
    string::const_iterator searchStart(corruptedMemory.cbegin());

    // Find all matches
    while (regex_search(searchStart, corruptedMemory.cend(), match, pattern)) {
        string fullMatch = match[1];

        if (fullMatch == "do()") {
            mulEnabled = true;
        } else if (fullMatch == "don't()") {
            mulEnabled = false;
        } else if (match[2].matched && match[3].matched) {
            // This is a mul(X,Y) instruction
            if (mulEnabled) {
                int num1 = stoi(match[2]);
                int num2 = stoi(match[3]);
                totalSum += num1 * num2;
            }
        }

        searchStart = match.suffix().first;
    }

    return totalSum;
}

int main() {
    // Read the corrupted memory from 'input.txt'
    ifstream inputFile("input.txt");
    if (!inputFile.is_open()) {
        cerr << "Error: Unable to open file input.txt" << endl;
        return 1;
    }

    string corruptedMemory((istreambuf_iterator<char>(inputFile)),
                           istreambuf_iterator<char>());
    inputFile.close();

    // Calculate the results
    int totalSumAllMulOperations = sumMulOperations(corruptedMemory);
    int totalSumEnabledMulOperations = sumEnabledMulOperations(corruptedMemory);

    // Output the results
    cout << "The sum of all mul operations is: " << totalSumAllMulOperations << endl;
    cout << "The sum of enabled mul operations is: " << totalSumEnabledMulOperations << endl;

    return 0;
}

File: other-programming-languages/PHP/2024/2024-003/main.php
<?php

function sumMulOperations($corruptedMemory) {
    // Regex pattern for mul(X,Y)
    $pattern = '/mul\((\d{1,3}),(\d{1,3})\)/';
    $total = 0;

    // Find all matches
    preg_match_all($pattern, $corruptedMemory, $matches, PREG_SET_ORDER);

    // Calculate the sum of products
    foreach ($matches as $match) {
        $num1 = intval($match[1]);
        $num2 = intval($match[2]);
        $total += $num1 * $num2;
    }

    return $total;
}

function sumEnabledMulOperations($corruptedMemory) {
    // Regex pattern for do(), don't(), and mul(X,Y)
    $pattern = '/(do\(\)|don\'t\(\)|mul\((\d{1,3}),(\d{1,3})\))/';
    $totalSum = 0;
    $mulEnabled = true; // mul instructions are enabled at the start

    // Find all matches
    preg_match_all($pattern, $corruptedMemory, $matches, PREG_SET_ORDER);

    foreach ($matches as $match) {
        $fullMatch = $match[1];

        if ($fullMatch === "do()") {
            $mulEnabled = true;
        } elseif ($fullMatch === "don't()") {
            $mulEnabled = false;
        } elseif (isset($match[2]) && isset($match[3])) {
            // This is a mul(X,Y) instruction
            if ($mulEnabled) {
                $num1 = intval($match[2]);
                $num2 = intval($match[3]);
                $totalSum += $num1 * $num2;
            }
        }
    }

    return $totalSum;
}

// Read the corrupted memory from 'input.txt'
$filePath = 'input.txt';
$corruptedMemory = file_get_contents($filePath);

// Calculate the results
$totalSumAllMulOperations = sumMulOperations($corruptedMemory);
$totalSumEnabledMulOperations = sumEnabledMulOperations($corruptedMemory);

// Output the results
echo "The sum of all mul operations is: $totalSumAllMulOperations\n";
echo "The sum of enabled mul operations is: $totalSumEnabledMulOperations\n";

?>
